### ISSUE TITLE
Range() Function does not overwrite local increment inside the for loop

### ISSUE THREAD

**Issue by faze-geek** [lcompilers/lpython#1514]
I came across an instance of incorrect behavior while using the range function in for loops -
```
def main():
    j: i32
    i: i32
    for i in range(5):
        j=i;
        print(j)
        i=i+1

main()

(lp) C:\Users\kunni\lpython>python try.py
0
1
2
3
4

(lp) C:\Users\kunni\lpython>src\bin\lpython try.py
0
2
4
```

The `i = i + 1` step does not affect the output because, in each iteration of the loop, the value of `i` is being overwritten by the next value generated by the range function. The range function generates a sequence of numbers from 0 to 4, and in each iteration of the loop, the value of `i `is set to the next value in the sequence.

**Comment by anutosh491** [comment]
Does this have something to do with the LLVM IR ? I mean to say if we compare (only the loop body section is of interest) what we have a simple for loop and a loop like above with incrementation , we have the following
```
def main():
    i: i32
    for i in range(5):
        print(i)

loop.body:                                        ; preds = %loop.head
  %3 = load i32, i32* %i, align 4
  %4 = add i32 %3, 1
  store i32 %4, i32* %i, align 4
  %5 = load i32, i32* %i, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @2, i32 0, i32 0), i32 %5, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  br label %loop.head
```
```
def main():
    i: i32
    for i in range(5):
        print(i)
        i=i+1

loop.body:                                        ; preds = %loop.head
  %3 = load i32, i32* %i, align 4
  %4 = add i32 %3, 1
  store i32 %4, i32* %i, align 4
  %5 = load i32, i32* %i, align 4
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([5 x i8], [5 x i8]* @2, i32 0, i32 0), i32 %5, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  %6 = load i32, i32* %i, align 4
  %7 = add i32 %6, 1
  store i32 %7, i32* %i, align 4
  br label %loop.head
```
The last 3-4 lines here are responsible for the incrementation of (`i += 1`) , which is not wrong as we might have to use `i` again in the loop body , but then shouldn't the last line of a loop body every for loop end with something like 
`store i32 %4, i32* %i, align 4` to set the value of i to be 1 more after every iteration . 
Is this the correct way to approach this @Thirumalai-Shaktivel @certik ?

**Comment by anutosh491** [comment]
I can see there is more than just the above case where we need to take care of local updates . For eg python doesn't allow locally updating an iterable if we're looping over it . 
```
def main():
    s: str = "abc"
    i: str
    for i in s:
        print(i)
        s = "xyz"

main()

(lf) anutosh491@spbhat68:~/lpython/lpython$ python examples/expr2.py 
a
b
c
(lf) anutosh491@spbhat68:~/lpython/lpython$ ./src/bin/lpython examples/expr2.py --show-stacktrace
a
y
z
```
So yeah only saving the state of i as I mentioned in the above comment ( through `store i32 %4, i32* %i, align 4`) might not be enough , we need to make sure that the iterables state is also preserved.

**Comment by faze-geek** [comment]
> The last 3-4 lines here are responsible for the incrementation of (`i += 1`) , which is not wrong as we might have to use `i` again in the loop body , but then shouldn't the last line of a loop body every for loop end with something like 
> `store i32 %4, i32* %i, align 4` to set the value of i to be 1 more after every iteration . 

I belive these 2 lines are responsible for incrementing the loop, so it is not incorrect .
```
%3 = load i32, i32* %i, align 4 
%4 = add i32 %3, 1
store i32 %4, i32* %i, align 4
```
If I'm not wrong ,
Here you are taking the content present at memory address location of i and loading it in register 3. Then adding 1 to value of register 3 and loading it into register 4. Then you are storing value of register 4 into memory address location of i .

**Comment by faze-geek** [comment]
@czgdp1807 Can you check the validity of this issue? I had a brief chat with Thirumalai on zulip who feels that the output is correct.

**Comment by czgdp1807** [comment]
IMO, the `lpython` output should match CPython output. So fixing this issue is the way forward.

**Comment by anutosh491** [comment]
I realize that we need to fix how the increment statement works in the `replace_doloop` function in `pass_utils.cpp`  .  https://github.com/lcompilers/lpython/blob/9c00b60cf034227b99c42b3c343dbf0db828dba3/src/libasr/pass/pass_utils.cpp#L724-L731

Here `inc_stmt` depends on `c` which is directly responsible for the incrementation happening in the for loop . But I guess that only handles the incrementation passed in `range` . If there's some local updation/incrementation , I guess they need to be reversed . For that we can save the initial state of the i and then reverse the local updation by pushing the reversal statements after this for loop
```
 for (size_t i=0; i<loop.n_body; i++) { 
     body.push_back(al, loop.m_body[i]); 
 } 
```
So basically , we might have to create something like this transformation 
```
for i in range(5):
        print(i)
        i+=1
```
Changes to something like 
```
for i in range(5):
        j = i
        print(i)
        i+=1
        i = j
```
or 
```
for i in range(5):
        print(i)
        i+=1
        i -= 1
```
I guess something along these lines should be done ?!

**Comment by Thirumalai-Shaktivel** [comment]
Yup, I think this might be correct.

**Comment by anutosh491** [comment]
> We should also ensure that everything in LFortran works as it did before

Sure, once I address this correctly . I'll open up a subsequent pr if needed in LFortran too !

**Comment by anutosh491** [comment]
Hello @Thirumalai-Shaktivel , I was reasonably convinced that the above shared diff , would do the job for us .... but I still get the wrong output for the following 
```
def main():
    i: i32
    for i in range(5):
        print(i)
        i=i+1

main()

(lf) anutosh491@spbhat68:~/lpython/lpython$ ./src/bin/lpython examples/expr2.py 
0
2
4
```
Am I missing something here ?

**Comment by Thirumalai-Shaktivel** [comment]
```diff
+                ASR::expr_t *dummy = target;
+                stmt3 = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, dummy, target, nullptr));
```
From above, I think you are just assigning the target value to the target .i.e., it becomes
```
i = i
```
Instead, you have to create a new variable using `ASR::make_Variable_t`,  make expr_t using ASR::Var_t (assign this result to the above `dummy`). I think that would work.

**Comment by anutosh491** [comment]
> Instead, you have to create a new variable using ASR::make_Variable_t, and enclose the Variable symbol within ASR::Var_t (it becomes expr_t, then you can assign this result to the above dummy). I think that would work.

I've tried to replicate these instructions .
```
                std::string name = "dummy";
                ASR::expr_t *dummy = ASR::down_cast<ASR::expr_t>(ASR::make_Var_t(
                    al, loc, ASR::down_cast<ASR::symbol_t>(ASR::make_Variable_t(
                        al, loc, nullptr, s2c(al, name), nullptr, 0, ASR::intentType::Local,
                        nullptr, nullptr, ASR::storage_typeType::Default,
                        ASRUtils::TYPE(ASR::make_Integer_t(al, loc, 4, nullptr, 0)),
                        ASR::abiType::Source, ASR::accessType::Public,
                        ASR::presenceType::Required, false))));
                stmt3 = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, dummy, target, nullptr));

```
Though , I'm not fully confident about some parameters in `make_Variable_t`, mainly the symbol table one .
```
    static inline asr_t* make_Variable_t(Allocator &al, const Location &a_loc, SymbolTable* a_parent_symtab, char* a_name, char** a_dependencies, size_t n_dependencies, intentType a_intent, expr_t* a_symbolic_value, expr_t* a_value, storage_typeType a_storage, ttype_t* a_type, abiType a_abi, accessType a_access, presenceType a_presence, bool a_value_attr) {
        Variable_t *n;
        n = al.make_new<Variable_t>();
        n->base.type = symbolType::Variable;
        n->base.base.type = asrType::symbol;
        n->base.base.loc = a_loc;
        n->m_parent_symtab = a_parent_symtab;
.........
........
```
I've passed a `nullptr` for `SymbolTable* a_parent_symtab` in the code above but maybe that's not what it should be . You could go through it once ? Also maybe have a look at other parameters too !

**Comment by anutosh491** [comment]
When I use the following diff as pasted above (changing nothing else other than how the dummy variable is defined ) , I get the following stack trace for the above program !
```
Traceback (most recent call last):
  Binary file "/home/anutosh491/lpython/lpython/src/bin/lpython", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  File "/home/anutosh491/lpython/lpython/src/bin/lpython.cpp", line 1682, in ??
    err = compile_python_to_object_file(arg_file, tmp_o, runtime_library_dir,
  File "/home/anutosh491/lpython/lpython/src/bin/lpython.cpp", line 705, in ??
    res = fe.get_llvm3(*asr, pass_manager, diagnostics, infile);
  File "/home/anutosh491/lpython/lpython/src/lpython/python_evaluator.cpp", line 58, in LCompilers::PythonCompiler::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    run_fn, infile);
  File "/home/anutosh491/lpython/lpython/src/libasr/codegen/asr_to_llvm.cpp", line 6836, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, pass_options, diagnostics);
  File "/home/anutosh491/lpython/lpython/src/libasr/pass/pass_manager.h", line 213, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/home/anutosh491/lpython/lpython/src/libasr/pass/pass_manager.h", line 106, in LCompilers::PassManager::_apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    if (!asr_verify(*asr, true, diagnostics)) {
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 795, in LCompilers::asr_verify(LCompilers::ASR::TranslationUnit_t const&, bool, LCompilers::diag::Diagnostics&)
    v.visit_TranslationUnit(unit);
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 119, in LCompilers::ASR::VerifyVisitor::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    this->visit_symbol(*a.second);
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4413, in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4157, in ??
    case symbolType::Function: { v.visit_Function((const Function_t &)x); return; }
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 338, in LCompilers::ASR::VerifyVisitor::visit_Function(LCompilers::ASR::Function_t const&)
    visit_stmt(*x.m_body[i]);
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4428, in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4206, in ??
    case stmtType::WhileLoop: { v.visit_WhileLoop((const WhileLoop_t &)x); return; }
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4952, in LCompilers::ASR::BaseWalkVisitor<LCompilers::ASR::VerifyVisitor>::visit_WhileLoop(LCompilers::ASR::WhileLoop_t const&)
    self().visit_stmt(*x.m_body[i]);
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4428, in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4178, in ??
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 274, in LCompilers::ASR::VerifyVisitor::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    BaseWalkVisitor<VerifyVisitor>::visit_Assignment(x);
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4708, in LCompilers::ASR::BaseWalkVisitor<LCompilers::ASR::VerifyVisitor>::visit_Assignment(LCompilers::ASR::Assignment_t const&)
    self().visit_expr(*x.m_target);
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4473, in LCompilers::ASR::BaseVisitor<LCompilers::ASR::VerifyVisitor>::visit_expr(LCompilers::ASR::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/home/anutosh491/lpython/lpython/src/libasr/../libasr/asr.h", line 4273, in ??
    case exprType::Var: { v.visit_Var((const Var_t &)x); return; }
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 579, in LCompilers::ASR::VerifyVisitor::visit_Var(LCompilers::ASR::Var_t const&)
    require(symtab_in_scope(current_symtab, x.m_v),
  File "/home/anutosh491/lpython/lpython/src/libasr/asr_verify.cpp", line 77, in LCompilers::ASR::VerifyVisitor::symtab_in_scope(LCompilers::SymbolTable const*, LCompilers::ASR::symbol_t const*)
    unsigned int symtab_ID = symbol_parent_symtab(sym)->counter;
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in killpg()
Segfault: Signal SIGSEGV (segmentation fault) received
```
Not exactly sure what's responsible for this !

**Comment by czgdp1807** [comment]
I think the following can be done,

```
for i in range(5):
    <line 1>
    <line 2>
    .
    .
    .
    <line n>
```

to

```
do i=1, 4
    j = i
   <line 1>
   <line 2>
   .
   .
   .
   <line n>
   i = j + 1
end do
```

`j` denotes a temporary variable in the current scope.

I think the above transformation should be done in AST to ASR directly. Though the semantics of `for...in` loop depend on the iterable. Basically `range` returns an iterable and `i` is set by shifting to the next element of the iterable, hence any changes to `i` inside the body of the loop is overwritten.

**Comment by Thirumalai-Shaktivel** [comment]
> https://github.com/lcompilers/lpython/issues/1514#issuecomment-1432814630

Here the parent_symtab is nullptr, but shouldn't be.
See: https://docs.lfortran.org/en/asr/asr_nodes/symbol_nodes/variable/
